### PR TITLE
Warning Hunt and CMake export.

### PR DESCRIPTION
**Pull request by MaximilienNaveau**
[//]: # "Thanks for your contribution.  To make life of the reviewers easier,"
[//]: # "please give this pull request a meaningful title and provide the"
[//]: # "requested information below."

## Description

[//]: # "Description of what you did.  If it is more complex, consider to add a"
[//]: # "'Summary' section."

Export the XXX.cmake file containing the targets and everything for the `find_package(master_board_sdk REQUIRED)`.
Also fix up a bunch of gcc warning about type conversion.

## How I Tested

[//]: # "Explain how you tested your changes"

WARNING not test yet, this is ongoing work.

### CODE REVIEW

File: sdk/master_board_sdk/CMakeLists.txt [hunk 1]
```diff
@@ -73,10 +71,14 @@ ADD_LIBRARY(${PROJECT_NAME} SHARED
   ${MASTER_BOARD_SRC}
 )
 TARGET_LINK_LIBRARIES(${PROJECT_NAME} pthread)
+target_include_directories(
+  ${PROJECT_NAME} PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
```
Comment (nim65s) [inline sdk/master_board_sdk/CMakeLists.txt:75]: I think the root `include` folder is automatically found by CMake, so I don't see the point with `$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>`. Do you ?

If you remove this, does it change anything ?

File: sdk/master_board_sdk/CMakeLists.txt [hunk 2]
```diff
@@ -73,10 +71,14 @@ ADD_LIBRARY(${PROJECT_NAME} SHARED
...
+                         $<INSTALL_INTERFACE:include>)
 install(
   TARGETS ${PROJECT_NAME}
   EXPORT ${TARGETS_EXPORT_NAME}
   DESTINATION lib)
+install(DIRECTORY include/ DESTINATION include)
```
Comment (nim65s) [inline sdk/master_board_sdk/CMakeLists.txt:81]: the JRL-CMakeModules is supposed to automatically installs includes. If you remove this directive, do you lack any header in the install prefix ?

File: sdk/master_board_sdk/CMakeLists.txt [hunk 3]
```diff
@@ -145,9 +147,12 @@ ENDIF(BUILD_PYTHON_INTERFACE)
 
 
 # --- CHECK EXAMPLES -----------------------------------------------------------
-ADD_EXECUTABLE(master-board-example example/example.cpp)
-TARGET_LINK_LIBRARIES(master-board-example ${PROJECT_NAME})
-INSTALL(TARGETS master-board-example DESTINATION bin)
+ADD_EXECUTABLE(master_board_example example/example.cpp)
+TARGET_LINK_LIBRARIES(master_board_example ${PROJECT_NAME})
+target_include_directories(
+    master_board_example PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
```
Comment (nim65s) [inline sdk/master_board_sdk/CMakeLists.txt:153]: Same thing here for `$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>`

File: sdk/master_board_sdk/package.xml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
+<package>
+    <name>master_board_sdk</name>
+    <version>1.0.0</version>
+    <description>
+        This project contains the sdk for the communication between a computer
+        and the master-board
+    </description>
+    <maintainer email="mnaveau@tue.mpg.de">Maximilien Naveau</maintainer>
+    <license>BSD 3-Clause</license>
```
Comment (nim65s) [inline sdk/master_board_sdk/package.xml:10]: I guess that's `BSD-3-Clause`

**Review by nim65s — COMMENTED** [review]
In general, this looks good to me, but I have a few questions

File: sdk/master_board_sdk/CMakeLists.txt (hunk 1 above)
Comment (MaximilienNaveau) [reply sdk/master_board_sdk/CMakeLists.txt:75]: I think it should be ok without it, it is just recommended by ROS2 ament. And it's modern cmake...
If I understood properly, it has better portability properties... I didn't dig too much about it I have to confess.
I do not have strong opinions on this. I can remove it.

File: sdk/master_board_sdk/CMakeLists.txt (hunk 2 above)
Comment (MaximilienNaveau) [reply sdk/master_board_sdk/CMakeLists.txt:81]: Without this instruction the includes are not installed properly.
Only the `include/master/board/sdk/` `config.hh`|`deprecated.hh`|`warning.hh`

File: sdk/master_board_sdk/CMakeLists.txt (hunk 3 above)
Comment (MaximilienNaveau) [reply sdk/master_board_sdk/CMakeLists.txt:153]: See above

File: sdk/master_board_sdk/package.xml (hunk 4 above)
Comment (MaximilienNaveau) [reply sdk/master_board_sdk/package.xml:10]: I have always seen it written `BSD 3-Clause`
https://tldrlegal.com/license/bsd-3-clause-license-(revised)
https://github.com/machines-in-motion/ubuntu_installation_scripts/blob/master/LICENSE
https://opensource.org/licenses/BSD-3-Clause (3-clause BSD ...)

Anyway I do not have strong opinion again...

File: sdk/master_board_sdk/CMakeLists.txt (hunk 1 above)
Comment (MaximilienNaveau) [reply sdk/master_board_sdk/CMakeLists.txt:75]: Ok so actually there is a CMake error if you do not use it:
```
CMake Error in CMakeLists.txt:
  Target "master_board_sdk" INTERFACE_INCLUDE_DIRECTORIES property contains
  path:

    "/home/mnaveau/devel/workspace_ros2/src/master-board/sdk/master_board_sdk/include"

  which is prefixed in the source directory.
```

File: sdk/master_board_sdk/package.xml (hunk 4 above)
Comment (nim65s) [reply sdk/master_board_sdk/package.xml:10]: Ok, I thought that http://wiki.ros.org/catkin/package.xml would recommend using the spdx-id, but it looks like you can put pretty much anything you want, thanks

Comment (MaximilienNaveau) [reply sdk/master_board_sdk/package.xml:10]: I did not know about this spdx-id. My bad. But indeed this field should be able to contain anything.
I think it is more for human to read than for some code parsing. So maybe no need for a norm here?
I do not know.

File: sdk/master_board_sdk/license.txt [hunk 5]
```diff
@@ -0,0 +1,29 @@
+BSD 3-Clause License
+
+Copyright (c) 2019, New York University and Max Planck Gesellschaft.
```
Comment (thomasfla) [inline sdk/master_board_sdk/license.txt]: Shouldn't LAAS-CNRS be mentioned ?

Comment (MaximilienNaveau) [reply sdk/master_board_sdk/license.txt]: Arg! sorry my bad!
I update this

Comment (MaximilienNaveau) [reply sdk/master_board_sdk/license.txt]: Acutally there is a license file already at the root of the master-board so I will delete this one

Comment (MaximilienNaveau) [reply sdk/master_board_sdk/license.txt]: What is `solo-bl` in https://github.com/open-dynamic-robot-initiative/master-board/blob/master/LICENSE#L3 ?

**Comment by MaximilienNaveau** [comment]
I pushed the last changes.
I wonder about the API, why do we use floats?
It is not very convenient as it flags warnings very often with gcc...

**Comment by thomasfla** [comment]
We use float here because this code was developed in the same time as the master board firmware witch is has a limited FPU.
On the PC side (on the sdk), we can totally switch to doubles. We just need to rewrite the conversion macros, and change the variable types.

I don't really like all the static_cast introduced by this PR. If switching to double remove the need for the cast, I think we should do it. What do you think?

**Comment by jviereck** [comment]
> I don't really like all the static_cast introduced by this PR. If switching to double remove the need for the cast, I think we should do it. What do you think?

I agree with Thomas here. What we could do is change the API to take doubles but internally work with floats.

**Comment by MaximilienNaveau** [comment]
That is exactly what I was thinking I will change the API while keeping the internal float so no need to rewrite macros (I do not feel like introducing stupid bugs because of it).

I will take care of this in this PR.

File: sdk/master_board_sdk/CMakeLists.txt [hunk 6]
```diff
@@ -66,89 +61,54 @@ ADD_LIBRARY(${PROJECT_NAME} SHARED
...
-ADD_EXECUTABLE(master-board-example example/example.cpp)
-TARGET_LINK_LIBRARIES(master-board-example ${PROJECT_NAME})
-INSTALL(TARGETS master-board-example DESTINATION bin)
+ADD_EXECUTABLE(master_board_example example/example.cpp)
+TARGET_LINK_LIBRARIES(master_board_example ${PROJECT_NAME})
+INSTALL(TARGETS master_board_example DESTINATION bin)
 
 ADD_EXECUTABLE(master-board-example-pd example/example_pd.cpp)
 TARGET_LINK_LIBRARIES(master-board-example-pd ${PROJECT_NAME})
 INSTALL(TARGETS master-board-example-pd DESTINATION bin)
```
Comment (jviereck) [inline sdk/master_board_sdk/CMakeLists.txt:114]: I guess these new targets should be changed to lower-case like `master_board_example_pd` as well?

File: sdk/master_board_sdk/example/example.cpp [hunk 7]
```diff
@@ -12,6 +12,12 @@
 
 int main(int argc, char **argv)
 {
+    if(argc != 2)
+    {
+        throw std::runtime_error("Please provide the interface name "
+                                 "(i.e. using 'ifconfig' on linux");
+    }
```
Comment (jviereck) [inline sdk/master_board_sdk/example/example.cpp:19]: Thanks for adding this!

**Review by jviereck — APPROVED** [review]

**Comment by MaximilienNaveau** [comment]
I guess this can be merged as well... May be someone can test it on a testbench if you really want?

**Comment by nim65s** [comment]
I'm late on the review, sorry, I'll try to make a final pass today

File: sdk/master_board_sdk/CMakeLists.txt (hunk 2 above)
Comment (nim65s) [reply sdk/master_board_sdk/CMakeLists.txt:71]: Ok, that's because they are not included in `${${PROJECT_NAME}_HEADERS}`

**Review by nim65s — APPROVED** [review]

**Comment by MaximilienNaveau** [comment]
@thomasfla  should I merge this?

**Review by thomasfla — APPROVED** [review]
Thanks @MaximilienNaveau

**Comment by thomasfla** [comment]
The code looks good to me, but I didn't tested it on hardware since I do not have an easy access to it at the moment.

**Pull request merged**